internal/zinc-core/src/main/scala/sbt/internal/inc/InvalidationProfiler.scala

```diff
@@ -0,0 +1,267 @@
+package sbt.internal.inc
+
+import java.io.File
+
+import xsbti.UseScope
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Defines a profiler interface that translates to the profiling protobuf schema.
+ *
+ * The protobuf schema has been mildly inspired from pprof:
+ * https://github.com/google/pprof/blob/master/proto/profile.proto
+ *
+ * A profiler interface should only be used by project, and not globally, as
+ * this class is not thread safe.
+ */
+abstract class InvalidationProfiler {
+  def profileRun: RunProfiler
+  def registerRun(run: zprof.ZincRun): Unit
+}
+
+object InvalidationProfiler {
+  final val empty: InvalidationProfiler = new InvalidationProfiler {
+    override def profileRun: RunProfiler = RunProfiler.empty
+    override def registerRun(run: zprof.ZincRun): Unit = ()
+  }
+}
+
+class ZincInvalidationProfiler extends InvalidationProfiler {
+  private final var lastKnownIndex: Long = -1L
+  private final val stringTable1: ArrayBuffer[String] = new ArrayBuffer[String](1000)
+  private final val stringTable2: ArrayBuffer[String] = new ArrayBuffer[String](10)
+  private final val stringTableIndices: mutable.HashMap[String, Long] =
+    new mutable.HashMap[String, Long]
+
+  def profileRun: RunProfiler = new ZincProfilerImplementation
+
+  private final var runs: List[zprof.ZincRun] = Nil
```

Comment:
Can you add some information why we need to persist every run here?

Also this list grows with every compile run if I understand this correctly?
AFAIK it's only used for serialization to disc. Should we empty the runs after
serialization? Or is the memory footprint  small enough that we should never care

Reply:
The `ZincProfiler` is managed outside of Zinc, by another tool. Whomever uses this profiler is responsible for managing the lifetime of the profilers used. That being said, the memory profile of the runs is very small given that we use a string table